ffi/src/ffi_tracing.rs

```diff
@@ -0,0 +1,354 @@
+                None => KernelStringSlice {
+                    ptr: std::ptr::null(),
+                    len: 0,
+                },
+            };
+            let event = Event {
+                message: msg,
+                level,
+                target,
+                line: metadata.line().unwrap_or(0),
+                file,
+            };
+            (self.callback)(event);
+        }
+    }
+}
+
+fn setup_event_subscriber(callback: TracingEventFn, max_level: Level) {
+    use tracing_subscriber::{layer::SubscriberExt, registry::Registry};
+    let filter: LevelFilter = max_level.into();
+    let event_layer = EventLayer { callback }.with_filter(filter);
+    let subscriber = Registry::default().with(event_layer);
+    tracing::subscriber::set_global_default(subscriber).expect("Failed to set global subscriber");
+}
+
+// utility code below for setting up the tracing subscriber for log lines
+
+type SharedBuffer = Arc<Mutex<Vec<u8>>>;
+
+struct TriggerLayer {
+    buf: SharedBuffer,
+    callback: TracingLogLineFn,
+}
+
+impl<S> Layer<S> for TriggerLayer
+where
+    S: Subscriber + for<'a> LookupSpan<'a>,
+{
+    fn on_event(&self, _: &TracingEvent<'_>, _context: Context<'_, S>) {
+        let mut buf = self.buf.lock().unwrap();
+        let output = String::from_utf8_lossy(&buf);
+        let message = kernel_string_slice!(output);
+        (self.callback)(message);
+        buf.clear();
+    }
+}
+
+#[derive(Default)]
+struct BufferedMessageWriter {
+    current_buffer: SharedBuffer,
+}
+
+impl io::Write for BufferedMessageWriter {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.current_buffer.lock().unwrap().extend_from_slice(buf);
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
+
+impl<'a> MakeWriter<'a> for BufferedMessageWriter {
+    type Writer = Self;
+
+    fn make_writer(&'a self) -> Self::Writer {
+        BufferedMessageWriter {
+            current_buffer: self.current_buffer.clone(),
+        }
+    }
+}
+
+fn setup_log_line_subscriber(
+    callback: TracingLogLineFn,
+    max_level: Level,
+    format: LogLineFormat,
+    ansi: bool,
+    with_time: bool,
+    with_level: bool,
+    with_target: bool,
+) {
+    use tracing_subscriber::{layer::SubscriberExt, registry::Registry};
+    let buffer = Arc::new(Mutex::new(vec![]));
+    let writer = BufferedMessageWriter {
+        current_buffer: buffer.clone(),
+    };
+    let fmt_layer = tracing_subscriber::fmt::layer()
+        .with_writer(writer)
+        .with_ansi(ansi)
+        .with_level(with_level)
+        .with_target(with_target);
+    let filter: LevelFilter = max_level.into();
+    let tracking_layer = TriggerLayer {
+        buf: buffer.clone(),
+        callback,
+    };
+
+    // This repeats some code, but avoids some insane generic wrangling if we try to abstract the
+    // type of `fmt_layer` over the formatter
```

Comment:
It's also when we call things like `without_time` which effects the second generic for `Format`. 

The real issue is, we want to return this. Really we just need an `impl Subscriber`, but because the concrete types all differ it gets really messy.

I can play around a bit more, but this felt overall more understandable and actually cleaner

Reply:
Also note that the `tracing_core` crate introduces `Dispatch` exactly to abstract over all this where they erase the type. But we can't do that so easily because we want to retain methods like `with_filter` and `pretty()`, etc.